database/database.go

```diff
@@ -208,3 +211,65 @@ func ensureCollection(ctx context.Context, db *mongo.Database, collName string)
 	}
 	return coll, nil
 }
+
+// generateUploadsDownloadsPipeline generates a mongo pipeline for transforming
+// an `Upload` or `Download` struct into the respective
+// `<Up/Down>loadResponseDTO` struct.
+//
+// The Mongo query we want to ultimately execute is:
+//	db.downloads.aggregate([
+//		{ $match: { "user_id": ObjectId("5fda32ef6e0aba5d16c0d550") }},
+//		{ $skip: 1 },
+//		{ $limit: 5 },
+//		{ $lookup: {
+//				from: "skylinks",
+//				localField: "skylink_id",  // field in the downloads collection
+//				foreignField: "_id",	   // field in the skylinks collection
+//				as: "fromSkylinks"
+//		  }
+//		},
+//		{ $replaceRoot: { newRoot: { $mergeObjects: [ { $arrayElemAt: [ "$fromSkylinks", 0 ] }, "$$ROOT" ] } } },
+//		{ $project: { fromSkylinks: 0 } }
+//	])
+//
+// This query will get all downloads by the current user, skip $skip of them
+// and then fetch $limit of them, allowing us to paginate. It will then
+// join with the `skylinks` collection in order to fetch some additional
+// data about each download.
+func generateUploadsDownloadsPipeline(matchStage bson.D, offset, pageSize int) mongo.Pipeline {
+	skipStage := bson.D{{"$skip", offset}}
+	limitStage := bson.D{{"$limit", pageSize}}
+	lookupStage := bson.D{
+		{"$lookup", bson.D{
+			{"from", "skylinks"},
+			{"localField", "skylink_id"}, // field in the downloads collection
+			{"foreignField", "_id"},      // field in the skylinks collection
+			{"as", "fromSkylinks"},
+		}},
+	}
+	replaceStage := bson.D{
+		{"$replaceRoot", bson.D{
+			{"newRoot", bson.D{
+				{"$mergeObjects", bson.A{
+					bson.D{{"$arrayElemAt", bson.A{"$fromSkylinks", 0}}}, "$$ROOT"},
```

Comment:
The double `$$` is intended?

Reply:
Yes. No clue why, tbh. I am balancing the "dig deeper, so you understand why" with "how do I just...". :)